controllers/noderefutil/providerid.go

```diff
@@ -96,3 +97,11 @@ func (p *ProviderID) String() string {
 func (p *ProviderID) Validate() bool {
 	return p.CloudProvider() != "" && p.ID() != ""
 }
+
+// IndexKey returns a string concatenating the cloudProvider and the ID parts of the providerID.
+// E.g Format: cloudProvider://optional/segments/etc/id. Parsed: cloudProvider/id
+// This is useful to use the providerID as a reliable index between nodes and machines
+// as it guarantees the infra Providers contract.
+func (p *ProviderID) IndexKey() string {
+	return fmt.Sprintf("%s/%s", p.CloudProvider(), p.ID())
```

Comment:
Wondering if dropping segments could lead to duplicated indexKey; e.g. looking at AWS or GCE, is it possible to have the same machine id in different regions? 
Would it be more straight forward/secure if we index for the entire providerID?

Reply:
I was just trying to be consistent with existing assumptions, cloudProvider and ID is what's currently used to compare, see https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/noderefutil/providerid.go#L86

In a follow up I'd be in favour of changing the expectation for the cluster-api-providers to set exactly the same providerID the controller manager cloud-provider sets. 
We actually ensured this in AWS a while ago https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1693.
Though there might be other provider specific reasons why this is not possible I'm not aware of.

Reply:
Created this to track the discussion https://github.com/kubernetes-sigs/cluster-api/issues/4526

Reply:
> Wondering if dropping segments could lead to duplicated indexKey; e.g. looking at AWS or GCE, is it possible to have the same machine id in different regions?

AWS or GCE machine names should be globally unique. cc @randomvariable @sedefsavas 

In general though, the ProviderID's id part is intended to be globally unique

Reply:
Instance ID may not be unique across regions, but the provider ID does encode the AZ, so this should be fine.

Reply:
>Instance ID may not be unique across regions, but the provider ID does encode the AZ (e.g. eu-west-1a), so this should be fine.

That invalidates our equality check.
Please let's contain this discussion in this thread to avoid having it scattered https://github.com/kubernetes-sigs/cluster-api/issues/4526#issuecomment-827473195

Reply:
I'll resolve this so we keep the discussion in thread.